ws-server/src/server.rs

```diff
@@ -21,477 +21,201 @@
-	pub fn register_method(&self, method_name: String) -> Result<RegisteredMethod, Error> {
-		if !self.registered_methods.lock().insert(method_name.clone()) {
-			return Err(Error::MethodAlreadyRegistered(method_name));
-		}
+					let json = serde_json::to_string(&JsonRpcResponse { jsonrpc: TwoPointZero, id, result: sub_id })?;
 
-		log::trace!("[frontend]: register_method={}", method_name);
-		let (tx, rx) = mpsc::channel(32);
+					tx.send(json).map_err(Into::into)
+				}),
+			);
+		}
 
-		self.to_back
-			.unbounded_send(FrontToBack::RegisterMethod { name: method_name, handler: tx })
-			.map_err(|e| Error::Internal(e.into_send_error()))?;
+		{
+			let subscribers = subscribers.clone();
+			self.methods.insert(
+				unsubscribe_method_name,
+				Box::new(move |id, params, tx, conn| {
+					let [sub_id]: [u64; 1] = serde_json::from_str(params)?;
+
+					subscribers.lock().remove(&(conn, sub_id));
+
+					let json = serde_json::to_string(&JsonRpcResponse {
+						jsonrpc: TwoPointZero,
+						id,
+						result: SubscriptionId::Num(0),
+					})?;
+
+					tx.send(json).map_err(Into::into)
+				}),
+			);
+		}
 
-		Ok(RegisteredMethod { to_back: self.to_back.clone(), queries_rx: rx })
+		SubsciptionSink { method: subscribe_method_name, subscribers }
 	}
 
-	/// Registers a subscription towards the server.
-	///
-	/// Clients will then be able to call this method.
-	/// The returned object allows you to send out notifications.
-	///
-	/// Returns an error if one of the method names was already registered.
-	pub fn register_subscription(
-		&self,
-		subscribe_method_name: String,
-		unsubscribe_method_name: String,
-	) -> Result<RegisteredSubscription, Error> {
-		{
-			let mut registered_methods = self.registered_methods.lock();
+	/// Build the server
+	pub async fn start(self, addr: impl AsRef<str>) -> anyhow::Result<()> {
+		let addr = addr.as_ref();
+		let mut incoming = TcpListenerStream::new(TcpListener::bind(addr).await?);
+		let methods = Arc::new(self.methods);
+		let mut id = 0;
 
-			// NOTE: We optimise for the case where neither method is registered.
-			// This means that if the strings are equal this will be slower than just comparing the
-			// strings.
-			if !registered_methods.insert(subscribe_method_name.clone()) {
-				return Err(Error::MethodAlreadyRegistered(subscribe_method_name));
-			}
-			if !registered_methods.insert(unsubscribe_method_name.clone()) {
-				registered_methods.remove(&subscribe_method_name);
-				return Err(Error::MethodAlreadyRegistered(unsubscribe_method_name));
-			}
-		}
-
-		log::trace!(
-			"[frontend]: server register subscription: subscribe_method={}, unsubscribe_method={}",
-			subscribe_method_name,
-			unsubscribe_method_name
-		);
-		let unique_id = self.next_subscription_unique_id.fetch_add(1, atomic::Ordering::Relaxed);
+		while let Some(socket) = incoming.next().await {
+			if let Ok(socket) = socket {
+				socket.set_nodelay(true).unwrap();
```

Comment:
Should we really panic if `set_nodelay` fails`? 
I guess that is a question on configuration, fine with ignoring this for now